WIP: Sure Petcare new features various improvements

benleb opened the pull request:
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
* add support for all four lock states
* add basic support for Sure Petcare Hub
* add support for the Sure Petcare Feeders (thanks to @ashscott from the home assistant forum)
* Add Connectitvity and Battery Sensors for all supported devices
* Cleanup code and simplify some things


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
```yaml
surepetcare:
  username: "username@surepetcare.io"
  password: sure_password
  feeders: [12345, 67890]
  flaps: [13579]
  pets: [24680]
```

## Additional information
- This PR fixes or closes issue: fixes #31429
- This PR is related to issue: 
- Link to documentation pull request: TODO!

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

benleb committed:
add typing

benleb committed:
add typing

benleb committed:
100% battery_level is enough

benleb committed:
100% battery_level is enough

benleb committed:
human-friendly datetime

benleb committed:
human-friendly datetime

benleb committed:
better enum usage

benleb committed:
better enum usage

benleb committed:
add online and learning mode attrs

benleb committed:
add online and learning mode attrs

benleb committed:
use max two decimals in attrs

benleb committed:
use max two decimals in attrs

benleb committed:
use legacy style debug logging

benleb committed:
use legacy style debug logging

benleb committed:
remove str usage of enums

benleb committed:
remove str usage of enums

benleb committed:
add feeder

benleb committed:
add feeder

benleb committed:
add feeder and adapt to new surepy version

benleb committed:
add feeder and adapt to new surepy version

benleb committed:
use ProductID instead of ThingID

benleb committed:
use ProductID instead of ThingID

benleb committed:
various changes and improvements

benleb committed:
various changes and improvements

benleb committed:
add connectivity sensors for all devices & proper support for multiple hubs

benleb committed:
add connectivity sensors for all devices & proper support for multiple hubs

Adminiuga committed:
ZHA dependencies bump (#31295)

* ZHA dependencies bump.

* Bump bellows-homeassistant.

balloob committed:
Handle service calls that do not refer entity IDs (#31317)

bramkragten committed:
Updated frontend to 20200130.0 (#31318)

balloob committed:
Fix HTTP config serialization (#31319)

engrbm87 committed:
Rework Mikrotik device scanning following Unifi (#27484)

* rework device scanning, add tests

* update requirements and coverage

* fix description comments

* update tests, fix disabled entity updates

* rework device scanning, add tests

* update requirements and coverage

* fix description comments

* update tests, fix disabled entity updates

* update librouteros to 3.0.0

* fix sorting

* fix sorting 2

* revert to 2.3.0 as 3.0.0 requires code update

* fix requirements

* apply fixes

* fix tests

* update hub.py and fix tests

* fix test_hub_setup_failed

* rebased on dev and update librouteros to 3.0.0

* fixed test_config_flow

* fixed tests

* fix test_config_flow

balloob committed:
Guard Z-Wave light HS conversion on None (#31320)

balloob committed:
Fix wemo lights (#31323)

rickvdl committed:
Add opening and closing states to MQTT covers (#31259)

* Added support for the opening and closing states to MQTT covers

* Processed PR feedback on MQTT cover changes

* Add missing MQTT abbreviation to fix failing tests

* Fixed typo in MQTT abbreviations

* Added mqtt set cover position optimistic test

endor-force committed:
Bump pytrafikverket to 0.1.6.1 (#30697)

* Bumped version for pytrafikverket

* Updated version for pytrafikverket

* Updated version for pytrafikverket

raman325 committed:
Update media_player and add tests to qualify vizio integration for platinum quality score (#31187)

* add media player test and update media player logic to qualify vizio for platinum quality score

* add SCAN_INTERVAL and log message once when device goes from available to unavailable and vice versa

* move test constants into one file to avoid defining dupes in each test file

* move constant to shrink diff

* move pytest fixtures to conftest.py

* remove commented out code

* move unload test to test_init

* updates to tests and logging based on review

* bypass notification service

* add fixture so component setup makes it through config flow

* split failure tests into two

* fix setup and entity check for test_init and setup failures in test_media_player

* make domain references consistent across test files

* remove logging steps that were introduced to help debug

* move common patches out to new fixture and use config entry everywhere appropriate

* fix docstring

* add test for update options to increase code coverage

* add one more assert

* refactor test_media_player to move boiler plate logic out of each test into _test_init function

* final refactor of test_media_player to move repeat logic into separate function

* update docstrings

* refactor setup failure tests to move shared logic into private function

* fix last new functions code to use variable instead of static config variable

* remove trailing comma

* test that volume_step gets properly passed to Vizio volume function

* fix comment language

* assert with unittest.mock.call in _test_service and use config_entries.async_setup instead of config_entries.async_add

* replace config_entries.async_add with config_entries.async_setup everywhere

* simplify if statement for argument assertion

* fix logging based on style guide

* remove accidentally committed changes

* update scan interval to something more reasonable and add tests for availability changes

* change filter function to list comprehension, simplify log messages, remove default entity id from logs since it is user configurable, fix docstrings

Commit:
Register on HA stop event to gracefully shutdown HomematicIP Cloud connections (#31289)

* Register on HA stop event to gracefully shutdown HomematicIP Cloud connections

* fixes after review

* Fix lint

* switch to unload_entry

* Save listener

* Switch back to hap.async_reset()

Kane610 committed:
Axis - config flow use new helper functions (#31286)

* Make use of new config flow helpers
Simplify Axis entry config to work with config flow helpers

* Keep old device data for rollback purposes

Kane610 committed:
UniFi - Try to discover local controller (#31326)

* Its working

* Use "unifi" as default host if a controller can be found

* Fix tests

* Make a fixture of patching the discovery function

frenck committed:
Bump pytest to 5.3.5 (#31327)

homeassistant committed:
[ci skip] Translation update

pnbruckner committed:
Fix async bug in amcrest when registering services (#31334)

xtools-at committed:
Emulated Hue: changed fallback device-type to fix Alexa compatibility issues (#30013) (#31330)

* Emulated Hue: changed the reported fallback device-type to fix Alexa compatibility issues (#30013)

* Emulated Hue: updated tests (#30013)

balloob committed:
Guard for callbacks in service helper (#31339)

balloob committed:
Fix incorrect annotation async flock notify (#31342)

* Fix incorrect annotation async flock notify

* Update notify.py

* Update notify.py

Co-authored-by: Pascal Vizeli <pascal.vizeli@syshack.ch>

MartinHjelmare commented on homeassistant/components/surepetcare/binary_sensor.py:
```diff
@@ -44,65 +63,77 @@ class SurePetcareBinarySensor(BinarySensorDevice):
+        self._device_class = device_class
+
+        self._spc: SurePetcareAPI = spc
+        self._spc_data: Dict[str, Any] = self._spc.states[self._sure_type].get(self._id)
```
Does it matter that we'll replace the spc data in the spc instance on every update? Ie, that `self._spc_data` here might become stale.

MartinHjelmare commented on homeassistant/components/surepetcare/binary_sensor.py:
```diff
@@ -44,65 +63,77 @@ class SurePetcareBinarySensor(BinarySensorDevice):
+        self._spc_data: Dict[str, Any] = self._spc.states[self._sure_type].get(self._id)
+        self._state: Dict[str, Any] = {}
+
+        _LOGGER.debug("%s -> self._spc_data: %s", self._id, self._spc_data)
```
We don't want side effects in init method. Please remove or move the logging.

benleb replied:
Thats just for debugging as its WIP, will remove this (and probably other logging currently in the code, also the commented code) before removing the "WIP" :)

MartinHjelmare commented on homeassistant/components/surepetcare/binary_sensor.py:
```diff
@@ -44,65 +63,77 @@ class SurePetcareBinarySensor(BinarySensorDevice):
+    def unique_id(self: BinarySensorDevice) -> str:
         """Return an unique ID."""
-        return f"{self._spc.household_id}-{self._id}"
+        return f"spc-{self._spc_data['household_id']}-{self._id}"
```
Changing the unique_id is a breaking change that we should avoid. We don't need any words anyway in the id. Ids are enough.

We do need a type id per entity type if a platform adds more than one entity type per device.

benleb replied:
fixed :)

MartinHjelmare commented on homeassistant/components/surepetcare/binary_sensor.py:
```diff
@@ -157,20 +188,136 @@ def device_state_attributes(self):
+                }
+
+            except (KeyError, TypeError) as error:
+                _LOGGER.error(
```
There should be no side effects in entity state properties.

benleb replied:
done 👍

raman325 committed:
Remove Throttle on async_setup and bump pyvizio version (#31337)

kellerza committed:
Upgrade pysma, fix #27154 (#31346)

frenck committed:
Partially Revert "Deprecate hide_if_away from device_tracker (#30833) (#31348)

cclauss committed:
Fix typos found by codespell (#31243)

* Fix typos found by codespell

* Fix typos found by codespell

* codespell: Furture  ==> Future

* Update test_config_flow.py

* Update __init__.py

* Spellcheck: successfull  ==> successful

* Codespell: unsuccesful  ==> unsuccessful

* Codespell: cant  ==> can't

* Codespell: firware ==> firmware

* Codespell: mimick  ==> mimic

dmulcahey committed:
bump quirks (#31355)

Kane610 committed:
UniFi integration move to push messaging (#31086)

* Rewrite UniFi integration to use push messaging

* Add signalling for new clients/devices

* Update list of known wireless clients when we get events of them connecting

* Reconnection logic for websocket

* Fix failing tests

* Bump requirement to v12

* Add new tests

* Update homeassistant/components/unifi/controller.py

Co-Authored-By: Martin Hjelmare <marhje52@gmail.com>

Kane610 committed:
Axis - Use core to start component tests (#31328)

* Fix test according to Martins comment in 31286

* Use core features rather than integration specific

* Fix populate options test

balloob committed:
Fix wemo device types for lights (#31360)

ochlocracy committed:
Protect for unknown state attributes. (#31354)

escoand committed:
drop fritzdect (#31359)

inverse committed:
Let core resolve entity_id for lastfm from username (#31280)

* Sluggify user

* Simplify

* Remove unused import

balloob committed:
Update Hue data fetching (#31338)

* Refactor Hue Lights to use DataCoordinator

* Redo how Hue updates data

* Address comments

* Inherit from Entity and remove pylint disable

* Add tests for debounce

homeassistant committed:
[ci skip] Translation update

DeltaTangoLima committed:
Updated wemo lights fix for #31360 (#31369)

springstan committed:
Fix auto_bypass in alarmdecoder (#30961)

* Fix auto_bypass in alarmdecoder

* Address review comments: used dict[key] and renamed variable

* Use dict[key] for required or optional config keys with default values

austinmroczek committed:
Add totalconnect zones as binary sensors (#28712)

* Bump skybellpy to 0.4.0

* Bump skybellpy to 0.4.0 in requirements_all.txt

* Added extra states for STATE_ALARM_TRIGGERED to allow users to know if
it is a burglar or fire or carbon monoxide so automations can take
appropriate actions.  Updated TotalConnect component to handle these new
states.

* Fix const import

* Fix const import

* Fix const imports

* Bump total-connect-client to 0.26.

* Catch details of alarm trigger in state attributes.

Also bumps total_connect_client to 0.27.

* Change state_attributes() to device_state_attributes()

* Move totalconnect component toward being a multi-platform integration.  Bump total_connect_client to 0.28.

* add missing total-connect alarm state mappings

* Made recommended changes of MartinHjelmare at
https://github.com/home-assistant/home-assistant/pull/24427

* Update __init__.py

* Updates per MartinHjelmare comments

* flake8/pydocstyle fixes

* removed . at end of log message

* added blank line between logging and voluptuous

* more fixes

* Adding totalconnect zones as HA binary_sensors

* fix manifest.json

* flake8/pydocstyle fixes.  Added codeowner.

* Update formatting per @springstan guidance.

* Fixed pylint

* Add zone ID to log message for easier troubleshooting

* Account for bypassed zones in update()

* More status handling fixes.

* Fixed flake8 error

* Another attempt at black/isort fixes.

* Bump total-connect-client to 0.50.  Simplify code using new functions in
total-connect-client package instead of importing constants.  Run black
and isort.

* Fix manifest file

* Another manifest fix

* one more manifest fix

* more manifest changes.

* sync up

* fix indent

* one more pylint fix

* Hopefully the last pylint fix

* make variable names understandable

* create and fill dict in one step

* Fix name and attributes

* rename to logical variable in alarm_control_panel

* Remove location_name from alarm_control_panel attributes since it is
already the name of the alarm.

* Multiple fixes to improve code per @springstan suggestions

* Update homeassistant/components/totalconnect/binary_sensor.py

Co-Authored-By: springstan <46536646+springstan@users.noreply.github.com>

* Multiple changes per @MartinHjelmare review

* simplify alarm adding

* Fix binary_sensor.py is_on

Co-authored-by: springstan <46536646+springstan@users.noreply.github.com>

vlebourl committed:
Add Tahoma lock support (#31311)

* Added lock support for tahoma

* Fixed logging to conform with pylint W1201 and E1205.

* Implemented @springstan suggestions.

* Fixed a typo...

* Implemented @springstan suggestions (v2).

* Implemented @springstan suggestions (v3).

scop committed:
Run mypy through a pyenv/virtualenv enabler wrapper script (#30922)

Kane610 committed:
deCONZ - Add support for new switch type (#31362)

balloob committed:
Remove hour delay before checking for updates (#31368)

* Check for updates at startup

* Add 100% test coverage for update_coordinator

* Address comments

basnijholt committed:
Add play, pause, previous and next track to kef (#31373)

* KEF: add support for play, pause, next track, and previous track

* bump aiokef to 0.2.7

* run isort

* do not dynamically change the supported features

dcnielsen90 committed:
Replace unmaintained BraviaRC backend with new fork: (#31234)

BraviaRC is currently unmaintained for home-assistant. This commit
    swaps it out with a new fork of BraviaRC called python-bravia-tv.
    This captures all bug fixes from BraviaRC release 3.7 (previously
    linked backend) to right before BraviaRC breaks when used by home-assistant.
    The intent of forking is to be able to continue supporting home-assistant.
    This is not intended to be a one off solution; this new fork will have
    future updates and be maintain as needed.

    This initial commit of python-bravia-tv improves the import process,
    however overall preserves the original API.
    Other fixes include:

    * Fix set-volume slider
    * Better error handling
    * Increase input options

    Resolves: #26351, #30964
    See also: #12577, #14843, #17345, #18245

Kane610 committed:
deCONZ - Services normalize bridge id (#31378)

* Services should also make sure to normalize bridge id since users do not know to manage themselves

Kane610 committed:
deCONZ - Improve config flow logging (#31381)

Kane610 committed:
UniFi - Log better information than a backtrace when poor switch data is involved (#31382)

fierland committed:
Add temperature and humidity to xiaomi miio air quality monitor (#31287)

* Added new attributes to the Miio airquality monitor:
- temperature
- humidity

* updated Docstrings

* docstrings updated

springstan committed:
Revert "Bump alarmdecoder to 1.13.9 (#30303)" (#31385)

This reverts commit f11d39f8ebf281a2069dc9f01777869c59405be4.

fabaff committed:
Add PlatformNotReady

Commit:
Support multiple attachments in signal messenger integration (#31141)

* added support for multiple attachments to signal_messenger integration

* updated pysignalclirestapi version in requirements_all.txt

* reworked multiple attachments feature in signal_messenger integration

* stay backwards compatible by both allowing the "attachment" and
  the "attachments" attribute.

* reworked multiple attachments feature in signal_messenger integration

* stay backwards compatible by both allowing the "attachment" and
  the "attachments" attribute.

* small change in signal_messenger integration

* added deprecation warning for 'attachment' attribute

* small changes in signal_messenger integration

* use 'warning' instead of 'warn' when logging the warning message

* re-generated requirements_test_pre_commit.txt

* added tests for signal_messenger integration

* regenerated requirements_test_all.txt for signal_messenger integration

* added more signal_messenger tests

* remove signal_messenger integration files from .coveragerc

fabaff committed:
Upgrade importlib-metadata to 1.5.0 (#31390)

homeassistant committed:
[ci skip] Translation update

ochlocracy committed:
Filter int in fan speed_list when yielding RangeController in Alexa (#31375)

* Allow for int in fan speed_list.

* Test for int in fan speed_list.

* prevent yielding preset for int labels.

fabaff committed:
Upgrade sendgrid to 6.1.1 (#31394)

fabaff committed:
Upgrade holidays to 0.10.1 (#31392)

Commit:
deCONZ - Fix magic cube awake gesture (#31403)

fabaff committed:
Upgrade discord.py to 1.3.1 (#31391)

fabaff committed:
Upgrade praw to 6.5.1 (#31393)

vanbalken committed:
Add unique_id to essent sensors (#31408)

* Fix mix-up of sensor entities and their values

* Prevent multiple calls for the same meter

balloob committed:
Catch device not found in device automations (#31401)

fabaff committed:
Upgrade numpy to 1.18.1 (#31411)

danielperna84 committed:
Update pyhomematic  to 0.1.64 (#31406)

akasma74 committed:
Fix rflink commands containing equals sign (#31412)

* new lib verson available

* new rflink lib version

* new rflink lib version

bjornorri committed:
Add Salt Fiber Box device tracker (#30986)

* Add salt component

* Update files from development checklist

* Use warning log level when data cannot be retrieved

Co-Authored-By: springstan <46536646+springstan@users.noreply.github.com>

* Remove empty fields from manifest.json

* Remove default arguments for host and username

* Bump saltbox library version

* Refactor and improve error handling

* Dev checklist

* Fix linting errors

* Check for None and return empty list

* Log on debug level instead of info

* More compact None checks

* Remove redundant None check

* Return None on exception but store as empty list

* More compact syntax

Co-authored-by: springstan <46536646+springstan@users.noreply.github.com>

fabaff committed:
Upgrade alpha_vantage to 2.1.3 (#31388)

gurbyz committed:
Search specific train in Nederlandse Spoorwegen (#28898)

* Nederlandse Spoorwegen: search for specific trip

* Reformatting with Black

* Resolve pylint error

* Reformat with black.

DevSecNinja committed:
Enable SUPPORT_VOLUME_STEP (#31023)

Kane610 committed:
deCONZ - Library cleanup (#31410)

* Follow library changes

* Bump dependency to v70

* Fix test

frenck committed:
Fix device name Google Assistant when using aliases (#31416)

* Fix device name Google Assistant when using aliases

* Adjust cloud tests

bendavid committed:
always call set_volume with integer values (#31418)

xtools-at committed:
Emulated Hue + Alexa: Fix devices not discovered and error response (#30013 & #29899) (#31413)

* Revert "Emulated Hue: changed the reported fallback device-type to fix Alexa compatibility issues (#30013)"

This reverts commit ddc8d9e25c0c8fd4073c0c516de9fa096cceb9bc.

* Revert "Emulated Hue: updated tests (#30013)"

This reverts commit 90df461e752fd6ecc1dc65bae0eba17f26a82f5f.

* Emulated Hue + Alexa: changed the fallback device-type again to "Dimmable Light" (#30013) after collective debugging; fixed brightness for on/off-devices and scripts to prevent "device malfunction" response from Alexa (#29899)

* Emulated Hue + Alexa: lint (#30013, #29899)

balloob committed:
Add dump service to MQTT integration (#31370)

* Add dump service to MQTT integration

* Lint

balloob committed:
Fix service annotations (#31402)

* Fix service annotations

* Filter area_id from service data

* Fix services not accepting entities

* Typo

bachya committed:
Streamline SimpliSafe data and token management (#31324)

* Streamline SimpliSafe API usage

* Streamline SimpliSafe data and token management

* Correctly define self.systems

* Inline update method

homeassistant committed:
[ci skip] Translation update

springstan committed:
Replace cmp option with eq and order (#31423)

raman325 committed:
Bump pyvizio version and add additional device info (#31417)

* bump pyvizio version and add additional device info

* add patches for get_model and get_version

* change keywrod argument to positional for _test_service

balloob committed:
Pass correct config to updater (#31428)

vlebourl committed:
Add guard clause for discovery_info to tahoma platforms (#31434)

benleb committed:
remove "side effects"/exception catching in attribs

benleb committed:
remove "side effects"/exception catching in attribs

YarmoM committed:
Update NSAPI to 3.0.2 (#30971)

* Bump NSAPI version to 3.0.1

* Compatibility with NSAPI 3.0.1 response

* Removed commented code

* Obsolete setups receive an upgrade notification

* Bump NS-API to 3.0.2

* Assign platform values directly

* Removed obsolete config warning

* Improved reference to obsolete password

benleb committed:
correct unique ids, reorder classes

benleb committed:
correct unique ids, reorder classes

benleb committed:
move Flap class from binary_sensor to sensor and add a sensore base class

benleb committed:
move Flap class from binary_sensor to sensor and add a sensore base class

benleb committed:
comments cleanup, minor typing and logging fixes

benleb committed:
comments cleanup, minor typing and logging fixes

benleb committed:
remove commented code

benleb committed:
remove commented code

benleb committed:
remove commented code

benleb committed:
remove commented code

benleb committed:
Merge branch 'sure-petcare-improvements' of github.com:benleb/home-assistant into sure-petcare-improvements

benleb commented:
oh fucked up the PR, sorry, will make a new one...^^
